### PR TITLE
feat(Delegate): add ability to add delegate

### PR DESCRIPTION
**Pull request by enddynayn**
Add the ability to add a delegate. Delegate authorization
is require via a signed payload.

This allows users to cost-shift message sending by allowing
the delegate to incur the cost.

[issue-86](https://github.com/LibertyDSNP/mrc/issues/86)

### CODE REVIEW

File: pallets/msa/src/types.rs [hunk 1]
```diff
@@ -36,3 +36,44 @@ where
...
+pub struct AddDelegate {
+	pub authorized_msa_id: MessageSenderId,
+	pub permission: u8,
+}
+
+#[derive(TypeInfo, Debug, Clone, Copy, Decode, Encode, PartialEq, MaxEncodedLen, Eq)]
+pub struct Delegator(pub MessageSenderId);
+
+impl From<MessageSenderId> for Delegator {
+	fn from(t: MessageSenderId) -> Self {
```
Comment (aramikm) [inline pallets/msa/src/types.rs:55]: looks 💪

File: pallets/msa/src/lib.rs [hunk 2]
```diff
@@ -102,6 +119,36 @@ pub mod pallet {
 			Ok(())
 		}
 
+		#[pallet::weight(10_000)]
```
Comment (aramikm) [inline pallets/msa/src/lib.rs:122]: these extrinsics need benchmark but it can be done in another issue or PR

File: pallets/msa/src/lib.rs [hunk 3]
```diff
@@ -237,7 +313,6 @@ impl<T: Config> Pallet<T> {
 		Self::get_key_info(&key).map(|info| info.msa_id)
 	}
 
-	/// Fetches all the keys associated with a message Source Account
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:240]: undo this from rebase

File: pallets/msa/src/lib.rs [hunk 4]
```diff
@@ -102,6 +119,36 @@ pub mod pallet {
...
+				&delegator_key,
+				&delegate_key,
+				payload_authorized_msa_id,
+			)?;
+
+			let _ = Self::add_delegate(delegate_msa_id, delegator_msa_id)?;
+
+			Self::deposit_event(Event::DelegateAdded {
+				delegator: delegator_msa_id.into(),
+				delegate: delegate_msa_id.into(),
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:146]: maybe dont need into now

File: pallets/msa/src/lib.rs [hunk 5]
```diff
@@ -52,6 +51,18 @@ pub mod pallet {
 	#[pallet::getter(fn get_identifier)]
 	pub type MsaIdentifier<T> = StorageValue<_, MessageSenderId, ValueQuery>;
 
+	#[pallet::storage]
+	#[pallet::getter(fn get_delegate_info_of)]
+	pub type DelegateInfoOf<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:58]: maybe add a different hash here because of imbalance tree

**Review by aramikm — APPROVED** [review]
Looks good! Added some nits and a question

File: pallets/msa/src/types.rs [hunk 6]
```diff
@@ -16,3 +16,44 @@ pub struct KeyInfo<BlockNumber> {
...
+}
+
+#[derive(TypeInfo, Debug, Clone, Decode, Encode, PartialEq)]
+pub struct AddDelegate {
+	pub authorized_msa_id: MessageSenderId,
+	pub permission: u8,
+}
+
+#[derive(TypeInfo, Debug, Clone, Copy, Decode, Encode, PartialEq, MaxEncodedLen, Eq)]
+pub struct Delegator(pub MessageSenderId);
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/msa/src/types.rs:52]: The reason i am using a struct is for having a strong type. The compiler will know that the right type (delegate, delegator) is inserted in the params. I think it is very easy to introduce a bug where we accidentally put delegate where it should be a delegator in the params.

File: pallets/msa/src/lib.rs [hunk 7]
```diff
@@ -187,6 +249,20 @@ impl<T: Config> Pallet<T> {
...
+	pub fn add_delegate(delegate: Delegate, delegator: Delegator) -> DispatchResult {
+		DelegateInfoOf::<T>::try_mutate(delegate, delegator, |maybe_info| -> DispatchResult {
+			ensure!(maybe_info.take() == None, Error::<T>::DuplicateDelegate);
+
+			let info = DelegateInfo { permission: Default::default() };
+
+			*maybe_info = Some(info);
+
+			Ok(())
+		})?;
```
[the remark this answers is not in the document]
Comment (enddynayn) [reply pallets/msa/src/lib.rs:284]: You mean remove the spacing right?

**Pull request merged**